src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java

```diff
@@ -597,8 +621,12 @@ void handleUnsuback(int unsubackMessageId) {
   void handlePuback(int pubackMessageId) {
 
     synchronized (this.connection) {
-      if (this.publishCompleteHandler != null) {
-        this.publishCompleteHandler.handle(pubackMessageId);
+      if (((MqttPublishVariableHeader) qos1outbound.peek().variableHeader()).messageId() == pubackMessageId) {
```

Comment:
here the first problem comes. The structure you need is something hybrid ... 
It works like a queue, because of course you need to publish in the order the user is calling the publish method but the ack could be received out of order so `peek()` and `poll()` don't fit well here. Imagine that you send two PUBLISH messages with id = 1 and id = 2 but the you receive the PUBACK for id = 2 first.